File: tests/omp/declare_target3.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->

  !! Omp.declare_target [cVarDef "p"];
  !! Omp.end_declare_target [cTopFunDefAndDecl "init"];
     let tg_loop = [cFor_c ""] in 
  
  !! Omp.parallel_for tg_loop;
  !! Omp.target tg_loop;
  !! Omp.target_update ~clause:[To_c ["v1";"v2"]] tg_loop;

  !! Omp.target_update ~clause:[From_c ["p"]] [cFun "output"];

)

File: tests/omp/declare_target3.cpp
const int N =1000;
float p[N], v1[N], v2[N];
extern void init(float *, float *, int);
extern void output(float *, int);
void vec_mult()
{
   int i;
   init(v1, v2, N);
   for (i=0; i<N; i++)
     p[i] = v1[i] * v2[i];
   output(p, N);
}
